src/main/java/com/bpodgursky/jbool_expressions/Expression.java

```diff
@@ -6,21 +6,14 @@
 
 public abstract class Expression<K> implements Comparable<Expression> {
 
-  public long hash;
-
   public int compareTo(Expression o) {
-    return Long.compare(this.hash, o.hash);
-  }
-
-  @Override
-  public boolean equals(Object o){
-    return o instanceof Expression &&
-           this.hash == ((Expression) o).hash;
-  }
-
-  @Override
-  public int hashCode() {
-    return (int) this.hash;
+    int compare = Integer.compare(hashCode(), o.hashCode());
```

Comment:
The advantage here is that the new code is much more Java and integrates naturally with objects that override hashCode appropriately.  The bad news is that this hashCode is not cached.  If this happens to be an expensive operation, it can be called many times in the simplification (even if we compute a hash at the NExpression level, we still look at the subexpressions in some of the Rules).  But I guess this seems quite unlikely.